File: project2/bisection.py
import numpy as np
"""
n = 10
A = np.zeros((n,n))
d = np.ones(n)*2
a =
a = -1
A[0,0] = d[0];
A[n-1,n-1] = d[n-1];
A[0,1] = A[n-1,n-2] = a;
for i in range(1,n-1):
    A[i,i] = d[i]
    A[i,i+1] = a
    A[i,i-1] = a

for i in range(1,n-1):
    xmin = d[i] - abs(a)
    xmax = d[i] + abs(a)
    for i in range(1,n):
"""
n = 5
c = np.ones(n)*2
b = np.ones(n)*(-1)
#c = np.zeros(n)
#b = np.zeros(n)
#c[0] = 1; c[1] = 49;
#b[1] = 7;
beta = b**2
m1 = 0; #m1 = 0;
m2 = n-1; #m2 = 4;
eps1 = 10**(-10)
relfeh = 2**(-20)
x = np.zeros(int(m2-m1+1))

beta[0]=b[0]=0.0;

# Set initial xmin and xmax - theorem???
xmin = c[n-1] - abs(b[n-1]);
xmax = c[n-1] + abs(b[n-1]);
for i in range(0,n-1):
    h = abs(b[i]) + abs(b[i+1]);
    if c[i] + h > xmax:
        xmax = c[i] + h;
    if c[i] - h < xmin:
        xmin = c[i] - h;

# ????
if xmin + xmax > 0:
    eps2 = relfeh*xmax;
else:
    eps2 = relfeh*(-xmin);
# ?????

if eps1 <= 0:
    eps1 = eps2;
eps2 = 0.5*eps1 + 7*eps2 # ??????

wu = np.zeros(m2-m1+1);
x0 = xmax;
for i in range(m1,m2+1):
    x[i] = xmax;
    wu[i] = xmin;
z = 0;
for k in range(m1,m2,1):
    xu = xmin;
    for i in range(k,m1,-1):
        if xu < wu[i]:
            xu = wu[i]
    if x0 > x[k]:
        x0 = x[k]
    while x0 - xu > 2*relfeh*(abs(xu)+abs(x0)) + eps1 and z < 10**5:
        x1 = (xu + x0)/2;
        z += 1
        # Sturms sequence
        a = 0; q = 1;
        for i in range(0,n-1):
            if q != 0:
                q = c[i] - x1 -  beta[i]/q;
            else:
                q = c[i] - x1 - abs(b[i]/relfeh)
            if q < 0:
                a += 1;
        if a < k:
            if a < m1:
                xu = wu[m1] = x1;
            else:
                xu = wu[a+1] = x1;
                if x[a] > x1:
                    x[a] = x1;
        else:
            x0 = x1;
    x[k] = (x0 + xu)/2

print(x)
print(z)

File: project2/bisection.cpp
#include <iostream>
#include <fstream>
#include <iomanip>
#include <cmath>
#include <string>
#include <armadillo>
#include "time.h"

// use namespace for output and input
using namespace std;
using namespace arma;
// object for output files
ofstream ofile;

int main(int argc, char *argv[]){
  int N;
  string ofilename;

  if(argc <= 1){
    cout << "Two few arguments in " << argv[0] <<
    ". Read in name of output file and the number of mesh points." << endl;
    exit(1);
  }
  else{
    ofilename = argv[1]; // Output file name
    N = atoi(argv[2]); // Matrix dimension
  }

  // Defining the elements of the tridiagonal matrix
  vec d(N, fill::zeros); // Vector of diagonal elements
  vec a(N, fill::zeros); // Vector of off-diagonal elements
  vec aa(N, fill::zeros); // Vector of off-diagonal elements squared
  for (int i = 0; i < N; i++){
    d(i) = 2;
    a(i) = -1;
    aa(i) = a(i)*a(i);
  }
  a(0) = aa(0) = 0; // Setting the first elements of a and a^2 to zero
  double eps, nonzero;
  int m1,m2;
  eps = pow (10.,-10); // The precision to which we want to know our lambda
  nonzero = pow (10.,-308); // The smallest number on the computer not zero
  m1 = 0; // The index of the lowest eigenvalue we want to find
  m2 = N-1; // The index of the largest eigenvalue we want to find

  // Setting up vectors to store information
  vec x(N, fill:: zeros); // Vector to store eigenvalues/upper bounds
  vec lb(N, fill::zeros); // Vector to store lower bounds for eigenvalues

  // Setting initial search interval xmin and xmax
  double xmax, xmin;
  xmin = d(N-1) - abs(a(N-1));
  xmax = d(N-1) + abs(a(N-1));
  for (int i = N-2; i > 0; i--){
    double h = abs(a(i)) + abs(a(i+1));
    if (d(i) + h > xmax){
      xmax = d(i) + h;
    }
    if (d[i] - h < xmax){
      xmin = d(i) - h;
    }
  }
  // Setting lower bound and upper bound vectors equal to xmin and xmax
  for (int i = m1; i < m2+1; i++){
    x(i) = xmax;
    lb(i) = xmin;
  }

  double xl; // Declaring lower bound x, xl
  double xu; // Declaring upper bound x, xu
  double xm; // Declaring middle x, xm
  double r; // Declaring ratio between polynomials
  int z, c; // Declaring counter of iterations z and counter of roots c

  xu = xmax;
  z = 0;

  clock_t start, finish; // declaring start and finish time
  start = clock(); // start time of computing

  for (int k = m2; k > m1-1; k--){ // Looping over all eigenvalues k

    xl = xmin;

    for (int i = k; i > m1-1; i--){
      if (xl < lb(i)){
        xl = lb(i); // Setting the lower bound to the largest lower bound
      }
      if (xu > x(k)){
        xu = x(k); // Setting the upper bound to the smallest upper bound
      }
    }

    while (xu - xl > eps && z < (int) pow(10,5)){

      xm = (xl + xu)/2.; // Setting middle point
      z += 1;

      // Sturm sequence
      c = 0; r = 1; // We start with r = 1 since P_0 = 1
      for (int i = 0; i < N; i++){
        if (r != 0){
          r = d(i) - xm - aa(i)/r;
        }
        else{
          r = d(i) - xm - aa(i)/nonzero; // avoid division by zero
        }
        if (r < 0){
          c += 1; // Counting roots
        }
      }

      // If the k-th root is larger than xm, set lower bound to xm
      if (c < k + 1){
        if (c < m1 + 1){  // If there are less roots than lowest number of roots we want to find
          xl = lb(m1) = xm; // - make xm the lowest bound for all eigenvalues we want to find
        }
        else{ // If not
          xl = lb(c) = xm; // - make xm the lower bound for the eigenvalue with index larger than the number of roots
          if (x(c-1) > xm){
            x(c-1) = xm; // make xm the upper bound for the eigenvalue with index lower than the number of roots
          }
        }
      }

      // If the k-th root is smaller than xm, et upper bound to xm
      else{
        xu = xm;
      }

    }
    x(k) = (xl + xu)/2.; // Set k-th eigenvalue (or the (k-1)-th upper bound) to xm
  }
  finish = clock(); // finish time of computing
  double time = ((double)(finish - start)/CLOCKS_PER_SEC);
  cout << setprecision(32) << "Time used for " << N << " gridpoints: "<< time << endl;
  x.print();
  lb.print();
  cout << z << endl;
}
